### ISSUE TITLE
TopK is incredibly slower than Pytorch on GPU

### ISSUE THREAD

**Issue by TianzhongSong** [microsoft/onnxruntime#2434]
**Describe the bug**
I have noticed https://github.com/microsoft/onnxruntime/issues/1830, and built onnxruntime from the latest source code. 
All nodes are placed on GPU, but the speed of TopK is still 15~20x slower than pytorch on GPU.

**Urgency**
None

**System information**
- OS Platform and Distribution: Ubuntu 18.04
- ONNX Runtime installed from: source
- ONNX Runtime version: 1.0.0
- Python version: 3.6
- Visual Studio version (if applicable):
- GCC/Compiler version (if compiling from source): 7.4.0
- CUDA/cuDNN version: CUDA 10.1 / cuDNN 7.6
- GPU model and memory: RTX 2080Ti / 11G

**To Reproduce**
pytorch script:

    import torch
    import torch.nn as nn
    from time import time

    class Model(nn.Module):
        def __init__(self):
            super(Model, self).__init__()
        
            self.net = nn.Sequential(
                nn.Linear(32, 64),
                nn.ReLU(inplace=True),
                nn.Linear(64, 128),
                nn.Softmax(dim=1)
            )

        def forward(self, x):
            x = self.net(x)
            x, _ = x.topk(5, dim=1)
            return x

    if __name__ == "__main__":
        model = Model()
        model.eval()

        input_tensor = torch.randn(16, 32)
        input_tensor = input_tensor.cuda()
        model = model.cuda()

        # warm up
        _ = model(input_tensor)

        t0 = time()
        for i in range(100):
            _ = model(input_tensor)
        t1 = time()
        print('PTH time: %f ms' %((t1-t0)*10))

        print('Export to onnx..')
        torch.onnx.export(model, input_tensor, "test_model.onnx", 
                        opset_version=11,
                        verbose=False, 
                        input_names=['input1'],
                        output_names=['output1'],
                        dynamic_axes={'input1':{0: 'batch'}, 'output1': {0: 'batch'}})

onnxruntime script:

    import onnxruntime as ort
    import numpy as np
    from time import time

    ort.set_default_logger_severity(0)
    print(ort.get_device())

    model = 'test_model.onnx'

    so = ort.SessionOptions()
    so.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL
    sess = ort.InferenceSession(model, so)

    inputs = np.empty((16, 32), dtype=np.float32)
    inputs = {'input1': inputs}

    # warm up
    for i in range(3):
        _ = sess.run(['output1'], inputs)

    t0 = time()
    for i in range(100):
        _ = sess.run(['output1'], inputs)
    t1 = time()
    print('ORT time: %f ms' %((t1-t0)*10))

**Expected behavior**

pytorch average forward time: 0.118899ms
onnxruntime average forward time: 0.390358ms

When remove `topk` operator, the forward speed of onnxruntime is faster than pytorch.

pytorch average forward time: 0.094895ms
onnxruntime average forward time: 0.051620ms

Obviously, onnxruntime's `TopK` is incredibly slower than pytorch's.

**Screenshots**
None

**Additional context**
Pytorch version: torch-1.4.0.dev20191118

**Comment by pranavsharma** [comment]
As stated in another [issue](https://github.com/microsoft/onnxruntime/issues/2404#issuecomment-554593653), have you ruled out the fact that your input is already on the GPU (```    input_tensor = input_tensor.cuda()```)?

**Comment by TianzhongSong** [comment]
@pranavsharma 

I have used [inference_session_test.cc](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/test/framework/inference_session_test.cc) to test my model, input and output are already on the GPU. Forward takes `378us`, which is similar with python demo (390us).

    1: [ RUN      ] InferenceSessionTests.TestIssue2434
    1: Elapsed time in microseconds : 378 µs
    1: Elapsed time in milliseconds : 0 ms

If remove the `topk` operator from model, forward only takes `26us`. 

    1: [ RUN      ] InferenceSessionTests.TestIssue2434
    1: Elapsed time in microseconds : 26 µs
    1: Elapsed time in milliseconds : 0 ms

**Comment by TianzhongSong** [comment]
Pytorch' `topk` only takes about 24us in this case.

**Comment by skottmckay** [comment]
How did you establish that all nodes were on GPU? There was no CUDA implementation of TopK in v1 of ONNX Runtime. It was added November 1, and with that I see a different issue that causes it to fail (looking into that, but it's a memory issue not performance once).

**Comment by TianzhongSong** [comment]
I tested this model with python api, and set `onnxruntime.set_default_logger_severity(0)`, output log as below:
```
2019-12-02 09:27:02.221911922 [V:onnxruntime:, inference_session.cc:442 TransformGraph] Node placements
2019-12-02 09:27:02.221916684 [V:onnxruntime:, inference_session.cc:444 TransformGraph] All nodes have been placed on [CUDAExecutionProvider].
2019-12-02 09:27:02.221946426 [I:onnxruntime:, session_state.cc:22 SetGraph] SaveMLValueNameIndexMapping
2019-12-02 09:27:02.221955692 [I:onnxruntime:, session_state.cc:67 SetGraph] Done saving OrtValue mappings.
```
It shows that all nodes have been placed on GPU.

**Comment by skottmckay** [comment]
https://github.com/microsoft/onnxruntime/pull/2523 will address the memory issue. Not sure if it will have a significant impact on performance.

The current TopK implementation does a full sort of the data. Something like https://github.com/facebook/fbcuda/blob/master/TopK.cuh could be considered. @RandySheriffH 

https://github.com/anilshanbhag/gpu-topk has some additional options but doesn't support key-value yet.

**Comment by RandySheriffH** [comment]
The fix might helps with perf already:
![image](https://user-images.githubusercontent.com/48490400/70094290-143e4a80-15d7-11ea-9807-2a86529232e2.png)
I am with GTX 1050 Ti on Linux. @TianzhongSong , would u mind try on your end?

**Comment by TianzhongSong** [comment]
Hi, i have tried again, but the speed of `topk` has no change.

**Comment by RandySheriffH** [comment]
Implementing http://anilshanbhag.in/static/papers/gputopk_sigmod18.pdf
let's if it helps.

**Comment by TianzhongSong** [comment]
@RandySheriffH Awesome job, speed of `topk` has been greatly improved. Thank you.

**Comment by TianzhongSong** [comment]
@RandySheriffH 

Returns wrong results when the size of the given dimension is too large.

Pytorch script:

    import torch
    import torch.nn as nn
    import numpy as np

    class Model(nn.Module):
        def __init__(self, k=3):
            super(Model, self).__init__()
            self.k = k

        def forward(self, x):
            v, ind = x.topk(self.k, dim=1)
            return v, ind

    model = Model()

    data = np.zeros((1, 1000), dtype=np.float32)
    data[0, 0:6] = np.array([0.1, 0.3, 0.02, 0.45, 0.08, 0.05])
    data = torch.from_numpy(data)

    score, index = model(data)
    print(score)
    print(index)

    torch.onnx.export(model, data, "topk.onnx",
                      opset_version=11,
                      verbose=False,
                      input_names=['data'],
                      output_names=['score', 'index'],
                      dynamic_axes={'data': {0: 'batch', 1: 'channel'},
                                    'score': {0: 'batch', 1: 'channel'},
                                    'index': {0: 'batch', 1: 'channel'}})

onnxruntime script:

    import onnxruntime as ort
    import numpy as np

    print(ort.get_device())

    model = 'topk.onnx'

    sess = ort.InferenceSession(model)

    data = np.zeros((1, 1000), dtype=np.float32)
    data[:, 0:6] = np.array([0.1, 0.3, 0.02, 0.45, 0.08, 0.05])

    inputs = {'data': data}

    score, index = sess.run(['score', 'index'], inputs)
    print(score)
    print(index)

Pytorch output:
    tensor([[0.4500, 0.3000, 0.1000]])
    tensor([[3, 1, 0]])

ORT output:
    [[0. 0. 0.]]
    [[0 0 0]]

Now, changing the size of dimension 1: `data = np.zeros((1, 500), dtype=np.float32)`, ORT output as below:
[[0.45 0.3  0.1 ]]
[[3 1 0]]

**Comment by TianzhongSong** [comment]
Issue has been fixed. Close this issue.

**Comment by Edwardmark** [comment]
@TianzhongSong How you fixed it?
